File: 04-1-one2nineapp_fl/lib/views/gameViewModel.dart

import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:one2nineapp_fl/game/gameConfig.dart';
import 'package:one2nineapp_fl/game/number.dart';
import 'package:one2nineapp_fl/persistence/score.dart';
import 'package:one2nineapp_fl/persistence/scoreRepository.dart';

class GameViewModel extends ChangeNotifier {
  // constant vars
  final gameConfig = GameConfig();
  final stopwatch = Stopwatch();
  final ScoreRepository repository;

  // constructor
  GameViewModel(this.repository);

  // other mutable vars
  int current = 1;
  List<Number> _numbers = [
    for (var i = 1; i <= 9; i++)
      Number(value: i, label: i.toString())
  ];
  List<bool> _marked = [false, false, false, false, false, false, false, false, false];

  bool showEndGameDialog = false;
  String endGameDialogMessage = "";
  double elapsedTime = 0.0;

  // getters
  get numbers => _numbers;
  get marked => _marked;

  void startGame() {
    gameConfig.reset();
    stopwatch.reset();
    stopwatch.start();
    startRound();
  }

  void startRound() {
    _marked = [false, false, false, false, false, false, false, false, false];
    _numbers = gameConfig.getNextConfiguration();
    current = 1;
    notifyListeners();
  }

  void click(int index, int value) {
    print("Clicked in index: $index, value of number: $value");
    if (value == current) {
      current++;
      _marked[index] = true;
    }

    if (current == 10) {
      if (gameConfig.hasNext()) {
        startRound();
      } else {
        endGame();
      }
    }

    notifyListeners();
  }

  void endGame() async {
    elapsedTime = stopwatch.elapsed.inMilliseconds / 1000.0;
    print("End game $elapsedTime");

    var bestScore = await repository.getBestScore();
    var isBest = "";
    if (bestScore == null || elapsedTime < bestScore.time) {
      isBest = "NEW RECORD!!!\n";
    }

    endGameDialogMessage = "${isBest}Your time was ${elapsedTime.toStringAsFixed(2)} seconds.";
    showEndGameDialog = true;
    notifyListeners();
  }

  void endDialog(String playerName) {
    showEndGameDialog = false;
    // save score
    if (playerName.trim().isEmpty) {
      playerName = "Anonymous";
    }
    DateFormat dateFormat = DateFormat("dd-MM-yyyy HH:mm:ss");
    String whenPlayed = dateFormat.format(DateTime.now());
    var newScore = Score(playerName, elapsedTime, whenPlayed);
    repository.insert(newScore);
  }
}

File: kotlin_projects/04-2-One2NineAppWithRoom_kt2/app/src/main/java/com/example/one2nineappwithroom_kt2/compose/GameViewModel.kt
package com.example.one2nineappwithroom_kt2.compose

import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.one2nineappwithroom_kt2.game.GameConfig
import com.example.one2nineappwithroom_kt2.game.Number
import com.example.one2nineappwithroom_kt2.repository.Score
import com.example.one2nineappwithroom_kt2.repository.ScoreRepository
import kotlinx.coroutines.launch
import java.lang.String
import java.text.SimpleDateFormat
import java.util.Date


class GameViewModel(private val scoreRepository: ScoreRepository) : ViewModel() {
    private val gameConfig = GameConfig()
    private var current = 1
    private var startTime : Long = 0
    private var yourTime = -1.0f
    var dialogTimeMessage = ""
    var markers = mutableStateListOf<Boolean>()
    var numbers = mutableStateListOf<Number>()
    var showEndGameDialog by mutableStateOf(false)
    var endGameDialogText by mutableStateOf("")
    var playerName by mutableStateOf("")

    init {
        startGame()
    }

    fun startGame() {
        gameConfig.reset()
        startTime = System.currentTimeMillis()
        startRound()
    }

    fun startRound() {
        current = 1
        markers.clear()
        repeat(9) {
            markers.add(false)
        }
        numbers.clear()
        gameConfig.getNextConfiguration().forEach { numbers.add(it) }
    }

    fun click(buttonIndex: Int, value: Int) {
        if (value == current) {
            // change to green
            current++
            markers[buttonIndex] = true
            Log.i("Game", "current: " + current)
            Log.i("Game", "" + markers.toString())
            if (current == 10) {
                if (gameConfig.hasNext()) {
                    startRound()
                } else {
                    endGame()
                }
            }
        }
    }

    fun endGame() {
        playerName = ""
        yourTime = (System.currentTimeMillis() - startTime) / 1000.0f
        viewModelScope.launch {
            val bestScore = scoreRepository.getBestScore()
            var isBest = ""
            if (bestScore == null || bestScore.time > yourTime)
                isBest = "NEW RECORD!!!\n"
            val message = isBest + String.format(dialogTimeMessage, yourTime)
            endGameDialogText = message
            showEndGameDialog = true
        }
    }

    fun closeDialog() {
        showEndGameDialog = false
        val whenPlayed = SimpleDateFormat("dd-MM-yyyy HH:mm:ss").format(Date())
        // save player data
        val player = playerName.takeIf { it.isNotEmpty() } ?: "Anonymous"
        val score = Score(
            playerName = player,
            time = yourTime,
            whenPlayed = whenPlayed)
        viewModelScope.launch {
            scoreRepository.insertScore(score)
        }
    }
}